### PR TITLE
make scaling factor a config arg of vae/vqvae

### PR DESCRIPTION
**Pull request by patil-suraj**
This PR makes the scaling factor a `config` argument of `vae` and `vqvae`.  The scaling factor is `0.18215` for all SD models except the [upscaler model](stabilityai/stable-diffusion-x4-upscaler), the upscaler VAE uses `0.08333`. 

This PR adds the argument `scaling_factor` and sets the default value to  `0.18215`  so all pipelines will work with it except `StableDiffusionUpscalePipeline`. To make it work for the `StableDiffusionUpscalePipeline`, we check if the `scaling_factor` is present in the `config` and if is set to `0.08333`, else it raises a deprecation warning and sets the value to `0.08333`. This is a bit hacky, but only the upscale pipeline is an exception to `0.08333`, and in this way we require fewer code changes. 

Another option would be to set the default value to `None`,  and then in every SD pipeline, check if the value is `None` and set it to an appropriate value, and then warn the user about this (like we are doing right now with the `sample_size`).

Let me know what you think @patrickvonplaten @pcuenca @anton-l.

### CODE REVIEW

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_upscale.py [hunk 1]
```diff
@@ -90,6 +92,25 @@ def __init__(
...
+                " version of any of the following: \n- stabilityai/stable-diffusion-x4-upscaler \n you should change"
+                " 'scaling_factor' to 0.08333 in the configuration file. Please make sure to update the config"
+                " accordingly as not setting `scaling_factor` in the config might lead to incorrect results in future"
+                " versions. If you have downloaded this checkpoint from the Hugging Face Hub, it would be very nice"
+                " if you could open a Pull request for the `vae/config.json` file"
+            )
+            deprecate("wrong scaling_factor", "1.0.0", deprecation_message, standard_warn=False)
+            new_config = dict(vae.config)
+            new_config["scaling_factor"] = 0.08333
+            vae._internal_dict = FrozenDict(new_config)
```
Comment (patil-suraj) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_upscale.py:112]: special case.

But actually thinking more about it, this way of adding a new argument, and deprecating in the same PR seems wrong.

Comment (pcuenca) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_upscale.py:111]: I think it should be ok as long as we add the parameter to `stabilityai/stable-diffusion-x4-upscaler` in the repo. Maybe we can open PRs for forked models too? Perhaps we can change the warning a little bit to clarify that the parameter might be absent from the json file and not set to a wrong value.

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_upscale.py [hunk 2]
```diff
@@ -90,6 +91,25 @@ def __init__(
...
+        )
+        if not is_vae_scaling_factor_set_to_0_08333:
+            deprecation_message = (
+                "The configuration file of the vae has  set the configuration `scaling_factor` to"
+                f" {vae.config.scaling_factor}. which seems highly unlikely. If your checkpoint is a fine-tuned"
+                " version of any of the following: \n- stabilityai/stable-diffusion-x4-upscaler \n you should change"
+                " 'scaling_factor' to 0.08333 in the configuration file. Please make sure to update the config"
+                " accordingly as not setting `scaling_factor` in the config might lead to incorrect results in future"
+                " versions. If you have downloaded this checkpoint from the Hugging Face Hub, it would be very nice"
+                " if you could open a Pull request for the `vae/config.json` file"
```
Comment (pcuenca) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_upscale.py:106]:
```suggestion
                "The configuration file of the vae does not contain `scaling_factor` or it is set to"
                f" {vae.config.scaling_factor}, which seems highly unlikely. If your checkpoint is a fine-tuned"
                " version of `stabilityai/stable-diffusion-x4-upscaler` you should change 'scaling_factor' to 0.08333"
                " Please make sure to update the config accordingly, as not doing so might lead to incorrect results"
                " in future versions. If you have downloaded this checkpoint from the Hugging Face Hub, it would be"
                " very nice if you could open a Pull request for the `vae/config.json` file"
```

**Review by pcuenca — APPROVED** [review]
Looks good to me!

File: src/diffusers/models/vae.py [hunk 3]
```diff
@@ -424,6 +424,7 @@ def __init__(
         num_vq_embeddings: int = 256,
         norm_num_groups: int = 32,
         vq_embed_dim: Optional[int] = None,
+        scaling_factor: int = 0.18215,
```
Comment (pcuenca) [inline src/diffusers/models/vae.py:427]: Maybe add a comment to the docstrings above, similar to what @Ghost---Shadow suggested in #1828: https://github.com/huggingface/diffusers/pull/1828/files

**Comment by patrickvonplaten** [comment]
This is a duplicate of https://github.com/huggingface/diffusers/pull/1515 - I should have linked it better from https://github.com/huggingface/diffusers/pull/1828.

In general, it's also good to check a bit if the PR exists, there was a longer discussion here: https://github.com/huggingface/diffusers/issues/726 .

Happy to close #1515 though in favor of this one.

Sorry @williamberman should have linked the issues/PRS from #1828 better.

**Comment by patil-suraj** [comment]
Aah, sorry didn't notice #1515 . Also in #1515 the scaling and unscaling are handled inside the `vae` model which IMO is not ideal, the `vae` itself does not require this, scaling is necessary when we are using the `vae` `latents`  as input to train some other model and the factor could be different depending on the dataset. So, IMO it should be done outside the model to make this explicit. WDYT?

**Comment by pcuenca** [comment]
@patil-suraj I agree, the vae itself does not require it. However, all diffusion pipelines use the latents and have to scale them, so I wouldn't be opposed to having special methods in the vae to achieve this (just not in the `encode` / `decode` path). But I'm not sure if something like `normalized_decode` would be any clearer than normalizing/denormalizing externally. One advantage of doing it inside the vae is that it'd be robust to more complex normalization in the future, but I don't think those are coming anytime soon.

File: src/diffusers/pipelines/stable_diffusion/pipeline_onnx_stable_diffusion.py [hunk 4]
```diff
@@ -289,7 +289,7 @@ def __call__(
             if callback is not None and i % callback_steps == 0:
                 callback(i, t, latents)
 
-        latents = 1 / 0.18215 * latents
+        latents = 1 / self.vae.config.scaling_factor * latents
```
Comment (anton-l) [inline src/diffusers/pipelines/stable_diffusion/pipeline_onnx_stable_diffusion.py:292]:
```suggestion
        latents = 1 / 0.18215 * latents
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_onnx_stable_diffusion_img2img.py [hunk 5]
```diff
@@ -353,7 +353,7 @@ def __call__(
         image = image.astype(latents_dtype)
         # encode the init image into latents and scale the latents
         init_latents = self.vae_encoder(sample=image)[0]
-        init_latents = 0.18215 * init_latents
+        init_latents = self.vae.config.scaling_factor * init_latents
```
Comment (anton-l) [inline src/diffusers/pipelines/stable_diffusion/pipeline_onnx_stable_diffusion_img2img.py:356]:
```suggestion
        init_latents = 0.18215 * init_latents
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_onnx_stable_diffusion_img2img.py [hunk 6]
```diff
@@ -436,7 +436,7 @@ def __call__(
             if callback is not None and i % callback_steps == 0:
                 callback(i, t, latents)
 
-        latents = 1 / 0.18215 * latents
+        latents = 1 / self.vae.config.scaling_factor * latents
```
Comment (anton-l) [inline src/diffusers/pipelines/stable_diffusion/pipeline_onnx_stable_diffusion_img2img.py:439]:
```suggestion
        latents = 1 / 0.18215 * latents
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_onnx_stable_diffusion_inpaint.py [hunk 7]
```diff
@@ -370,7 +370,7 @@ def __call__(
         masked_image = masked_image.astype(latents.dtype)
 
         masked_image_latents = self.vae_encoder(sample=masked_image)[0]
-        masked_image_latents = 0.18215 * masked_image_latents
+        masked_image_latents = self.vae.config.scaling_factor * masked_image_latents
```
Comment (anton-l) [inline src/diffusers/pipelines/stable_diffusion/pipeline_onnx_stable_diffusion_inpaint.py:373]:
```suggestion
        masked_image_latents = 0.18215 * masked_image_latents
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_onnx_stable_diffusion_inpaint.py [hunk 8]
```diff
@@ -443,7 +443,7 @@ def __call__(
             if callback is not None and i % callback_steps == 0:
                 callback(i, t, latents)
 
-        latents = 1 / 0.18215 * latents
+        latents = 1 / self.vae.config.scaling_factor * latents
```
Comment (anton-l) [inline src/diffusers/pipelines/stable_diffusion/pipeline_onnx_stable_diffusion_inpaint.py:446]:
```suggestion
        latents = 1 / 0.18215 * latents
```

**Comment by patil-suraj** [comment]
That's a good point @pcuenca, but another reason I choose to do it externally backward compatibility, there are a lot of custom scripts and pipelines that are doing this externally, and if we do scaling inside the model it could break those custom pipelines/scripts. 

Also, I think normalizing inputs is essentially a part of data processing which ideally should be separated from the model.

File: src/diffusers/pipelines/stable_diffusion/pipeline_onnx_stable_diffusion_inpaint_legacy.py [hunk 9]
```diff
@@ -347,7 +347,7 @@ def __call__(
 
         # encode the init image into latents and scale the latents
         init_latents = self.vae_encoder(sample=image)[0]
-        init_latents = 0.18215 * init_latents
+        init_latents = self.vae.config.scaling_factor * init_latents
```
Comment (anton-l) [inline src/diffusers/pipelines/stable_diffusion/pipeline_onnx_stable_diffusion_inpaint_legacy.py:350]:
```suggestion
        init_latents = 0.18215 * init_latents
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_onnx_stable_diffusion_inpaint_legacy.py [hunk 10]
```diff
@@ -426,7 +426,7 @@ def __call__(
             if callback is not None and i % callback_steps == 0:
                 callback(i, t, latents)
 
-        latents = 1 / 0.18215 * latents
+        latents = 1 / self.vae.config.scaling_factor * latents
```
Comment (anton-l) [inline src/diffusers/pipelines/stable_diffusion/pipeline_onnx_stable_diffusion_inpaint_legacy.py:429]:
```suggestion
        latents = 1 / 0.18215 * latents
```

**Review by anton-l — COMMENTED** [review]
Quick rollback: the ONNX models don't have a config, so the pipelines will have to rely on hardcoded values for now.

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_upscale.py (hunk 2 above)
Comment (patrickvonplaten) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_upscale.py:106]: Also for this could you please make a mass PR on the Hub?

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_upscale.py [hunk 11]
```diff
@@ -90,6 +91,25 @@ def __init__(
     ):
         super().__init__()
 
+        # check if vae has a config attribute `scaling_factor` and if it is set to 0.08333, else set it to 0.08333 and deprecate
+        is_vae_scaling_factor_set_to_0_08333 = (
+            hasattr(vae.config, "scaling_factor") and vae.config.scaling_factor == 0.08333
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_upscale.py:96]: this is a bit dangerous with `== and float numbers no?`

I would maybe just check if the vae has a scaling factor and if it does, we don't deprecate anything.

**Comment by patrickvonplaten** [comment]
Thanks for the PR @patil-suraj,

I think this PR follows the right approach. Just two things: 
- 1.) I would deprecate the upscaling pipeline a bit differently in that you only throw a deprecation warning (and set a value) if no `scale` value is defined
- 2.) Once this PR is merged we should do a mass PR (happy to send you a script for this)

**Comment by williamberman** [comment]
> This is a duplicate of #1515 - I should have linked it better from #1828.
> 
> In general, it's also good to check a bit if the PR exists, there was a longer discussion here: #726 .
> 
> Happy to close #1515 though in favor of this one.
> 
> Sorry @williamberman should have linked the issues/PRS from #1828 better.

No worries! Mine was stale and was definitely a wip. Happy to close out in favor of this :)

**Comment by williamberman** [comment]
> That's a good point @pcuenca, but another reason I choose to do it externally is backward compatibility. There are a lot of custom scripts and pipelines that are already doing this externally, and if we do scaling inside the model it could break those custom pipelines/scripts.
> 
> Also, I think normalizing inputs is essentially a part of data processing which ideally should be separated from the model.

I think handling inside the vae results in less code duplication. I.e. the scaling code only occurs in once place, and this would be backwards compatible with custom scripts and pipelines if the default behavior in the vae is to not do any scaling. 

Scaling by default in the pipeline because it requires fewer code changes also makes sense to me. I have no strong opinions here and am ok with either way :)

File: src/diffusers/models/vae.py (hunk 3 above)
Comment (williamberman) [reply src/diffusers/models/vae.py:427]: If helpful in docstring, originally documented in LDM sections 4.3.2 and D.1 https://arxiv.org/abs/2112.10752

**Comment by patrickvonplaten** [comment]
@patil-suraj the longer you wait the more merge-conflicts ;-)

**Comment by patil-suraj** [comment]
>1.) I would deprecate the upscaling pipeline a bit differently in that you only throw a deprecation warning (and set a value) if no scale value is defined

@patrickvonplaten  the upscale pipeline uses a different scale than other pipelines, so even if the value is present, we need to ensure it's the right value, as the default value is not used for the upscale model.

File: src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py [hunk 12]
```diff
@@ -224,7 +224,8 @@ def _generate(
         # Create init_latents
         init_latent_dist = self.vae.apply({"params": params["vae"]}, image, method=self.vae.encode).latent_dist
         init_latents = init_latent_dist.sample(key=prng_seed).transpose((0, 3, 1, 2))
-        init_latents = 0.18215 * init_latents
+        init_latents = self.vae.config.scaling_factor * init_latents
+        latents_shape = (batch_size, self.unet.in_channels, height // 8, width // 8)
```
Comment (pcuenca) [inline src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py:228]: Why is this shape computed here again? It was already used in lines 212 to 223.

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_upscale.py [hunk 13]
```diff
@@ -89,6 +89,22 @@ def __init__(
...
+        )
+        if not is_vae_scaling_factor_set_to_0_08333:
+            deprecation_message = (
+                "The configuration file of the vae does not contain `scaling_factor` or it is set to"
+                f" {vae.config.scaling_factor}, which seems highly unlikely. If your checkpoint is a fine-tuned"
+                " version of `stabilityai/stable-diffusion-x4-upscaler` you should change 'scaling_factor' to 0.08333"
+                " Please make sure to update the config accordingly, as not doing so might lead to incorrect results"
+                " in future versions. If you have downloaded this checkpoint from the Hugging Face Hub, it would be"
+                " very nice if you could open a Pull request for the `vae/config.json` file"
```
Comment (pcuenca) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_upscale.py:103]:
```suggestion
                " very nice if you could open a Pull Request for the `vae/config.json` file"
```

**Review by pcuenca — APPROVED** [review]

File: src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py (hunk 12 above)
Comment (patil-suraj) [reply src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py:228]: Good catch, maybe got leftover from merge conflicts.

File: src/diffusers/models/autoencoder_kl.py [hunk 14]
```diff
@@ -56,6 +56,13 @@ class AutoencoderKL(ModelMixin, ConfigMixin):
         act_fn (`str`, *optional*, defaults to `"silu"`): The activation function to use.
         latent_channels (`int`, *optional*, defaults to `4`): Number of channels in the latent space.
         sample_size (`int`, *optional*, defaults to `32`): TODO
+        scaling_factor (`float`, *optional*, defaults to `0.18215`):
```
Comment (patrickvonplaten) [inline src/diffusers/models/autoencoder_kl.py:59]: very nice docstring! 
```suggestion
        scaling_factor (`float`, *optional*, defaults to 0.18215):
```

File: src/diffusers/models/autoencoder_kl.py [hunk 15]
```diff
@@ -56,6 +56,13 @@ class AutoencoderKL(ModelMixin, ConfigMixin):
         act_fn (`str`, *optional*, defaults to `"silu"`): The activation function to use.
         latent_channels (`int`, *optional*, defaults to `4`): Number of channels in the latent space.
```
Comment (patrickvonplaten) [inline src/diffusers/models/autoencoder_kl.py:57]:
```suggestion
        latent_channels (`int`, *optional*, defaults to 4): Number of channels in the latent space.
```

Comment (patrickvonplaten) [reply src/diffusers/models/autoencoder_kl.py:57]: pure ints or floats we shouldn't put in `` I think

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_upscale.py (hunk 11 above)
Comment (patrickvonplaten) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_upscale.py:94]: Can you please also open a PR for the upscaling pipeline to make sure it's updated there?

**Review by patrickvonplaten — APPROVED** [review]
Thanks a lot!

File: src/diffusers/models/vae_flax.py [hunk 16]
```diff
@@ -754,6 +754,13 @@ class FlaxAutoencoderKL(nn.Module, FlaxModelMixin, ConfigMixin):
             Norm num group
         sample_size (:obj:`int`, *optional*, defaults to `32`):
             Sample input size
+        scaling_factor (`float`, *optional*, defaults to `0.18215`):
```
Comment (patrickvonplaten) [inline src/diffusers/models/vae_flax.py:757]:
```suggestion
        scaling_factor (`float`, *optional*, defaults to 0.18215):
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_upscale.py (hunk 11 above)
Comment (patrickvonplaten) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_upscale.py:94]: I mean open a PR here: https://huggingface.co/stabilityai/stable-diffusion-x4-upscaler/discussions
